apps/sr-frontend/src/app/Database.tsx

```diff
@@ -0,0 +1,148 @@
+
+export default function Database() {
+  const {
+    database: mainDB,
+    databases,
+    setDatabase,
+  } = useContext(DatabaseContext)
+  const [index, setIndex] = useState(0)
+  const database = databases[index]
+  const current = database === mainDB
+  const [data, setData] = useState('')
+  // Need to update the dbMeta when database changes
+  const [{ name, lastEdit }, setDBMeta] = useState(database.dbMeta.get())
+  useEffect(() => setDBMeta(database.dbMeta.get()), [database])
+
+  const { importedDatabase } =
+    useMemo(() => {
+      if (!data) return undefined
+      let parsed: any
+      try {
+        parsed = JSON.parse(data)
+        console.log(parsed)
+        if (typeof parsed !== 'object') {
+          return undefined
+        }
+      } catch (e) {
+        return undefined
+      }
+      // Figure out the file format
+      if (parsed.format === 'SROD' || parsed.format === 'SRO') {
+        // Parse as SROD format
+        const copyStorage = new SandboxStorage()
+        copyStorage.copyFrom(database.storage)
+        const importedDatabase = new SroDatabase(
+          (index + 1) as 1 | 2 | 3 | 4,
+          copyStorage
+        )
+        const importResult = importedDatabase.importSROD(parsed, false, false)
+        if (!importResult) {
+          return undefined
+        }
+
+        return { importResult, importedDatabase }
+      }
+      return undefined
+    }, [data, database, index]) ?? {}
+
+  const onUpload = async (e: ChangeEvent<HTMLInputElement>) => {
+    const file = (e.target.files ?? [''])[0]
+    if (typeof file === 'string') return
+    const reader = new FileReader()
+    reader.onload = () => setData(reader.result as string)
+    reader.readAsText(file)
+  }
+
+  const download = useCallback(() => {
+    const date = new Date()
+    const dateStr = date
+      .toISOString()
+      .split('.')[0]
+      .replace('T', '_')
+      .replaceAll(':', '-')
+    const JSONStr = JSON.stringify(database.exportSROD())
+    const filename = `${name.trim().replaceAll(' ', '_')}_${dateStr}.json`
+    const contentType = 'application/json;charset=utf-8'
+    const a = document.createElement('a')
+    a.download = filename
+    a.href = `data:${contentType},${encodeURIComponent(JSONStr)}`
+    a.target = '_blank'
+    document.body.appendChild(a)
+    a.click()
+    document.body.removeChild(a)
+  }, [database, name])
+
+  const replaceDb = useCallback(() => {
+    if (!importedDatabase) return
+    importedDatabase.swapStorage(database)
+    setDatabase(index, importedDatabase)
+    importedDatabase.toExtraLocalDB()
+  }, [database, importedDatabase, index, setDatabase])
+
+  const swapDb = useCallback(() => {
+    if (current) return
+    mainDB.toExtraLocalDB()
+    database.swapStorage(mainDB)
+    setDatabase(index, database)
+  }, [current, database, index, mainDB, setDatabase])
+
+  // TODO: Make this reload the component
+  const clearDb = useCallback(() => {
+    database.clear()
```

Comment:
@frzyc should this be triggering a re-render, or should I be manually triggering it myself?

Reply:
try it out, if clearing db doesnt update UI, then i'd say you prob have to either manual trigger or add a listener